add PDLL support to iree

IanWood1 opened the pull request:
Note: I messed up merging with main and 5 extra people got added automatically as reviewers, so I fixed the branch and recreated the old PR (https://github.com/iree-org/iree/pull/17097).

### Overview
[PDL Language (PDLL)](https://mlir.llvm.org/docs/PDLL/) is a declarative frontend language for creating pattern rewrites for MLIR. PDLL provides a intuitive interface for creating rewrites. **Patterns** allow users to declare a sequence of ops to match against and describe how they should be rewritten. PDLL allows for **Constraints** and **Rewriters** to either be written within PDLL or written in C++ and exposed through the C++ API. This PR contains examples of both native constraints and native rewriters within the mlp_plugin folder.



`mlir-pdll` provides functionality for parsing PDLL and generating either C++ or [PDL](https://mlir.llvm.org/docs/Dialects/PDLOps/). Adding support for PDLL requires the following functionality from `mlir-pdll`:
1. Generating a PDLL AST from the input file (`parsePDLLAST()`)
2. Generating a PDL pattern module from the PDLL AST (`codegenPDLLToMLIR()`)

After the PDL module has been constructed, the implementation is nearly identical to `ApplyPDLPatterns.cpp`, which is already added to iree to support PDL patterns.


### PR Summary:
- Added PDLL support to iree via the `iree-preprocessing-pdll-spec-filename` flag (matching `iree-preprocessing-pdl-spec-filename` already in place for PDL). This flag allows iree to use a pdll spec in the preprocessing pass.
- Added tests to tosa plugins, torch plugins, and mlp_plugins 
- Small update to extensions.md mentioning the new flag
- Added small readme in mlp_plugins that lists the 3 different languages used (transform dialect, pdl, pdll)

---

Since I had to create a new PR here are the requested changes from the old PR:
1. Removed errant file (https://github.com/iree-org/iree/pull/17097#discussion_r1575671594)
3. Removed unnecessary include (https://github.com/iree-org/iree/pull/17097#discussion_r1575672705)
4. Moved PDLL preprocessing pass out of the `ApplyPDLPatterns.cpp` file and into the new `ApplyPDLLPatterns.cpp`. Since PDLL relied on functions in the PDL file, I moved the utility functions to `Utilities.cpp/h` (https://github.com/iree-org/iree/pull/17097#discussion_r1575673162).
5. Created a torch test (https://github.com/iree-org/iree/pull/17097#discussion_r1576364766) 

---

### byo llvm related stuff

[Stella's work on mlir-tblgen](https://reviews.llvm.org/D153393)
[PDLL Parser cmake](https://github.com/llvm/llvm-project/blob/ddb67e68474092435634dd8a8e1c166f6ecd0e1a/mlir/lib/Tools/PDLL/Parser/CMakeLists.txt)
[mlir-tblgen lib cmake](https://github.com/llvm/llvm-project/blob/ddb67e68474092435634dd8a8e1c166f6ecd0e1a/mlir/lib/Tools/mlir-tblgen/CMakeLists.txt)
[lib tablegen cmake](https://github.com/llvm/llvm-project/blob/6f02120ac4463e5e0cda25e2aafc485a4fe634ea/mlir/lib/TableGen/CMakeLists.txt)

IanWood1 committed:
squash all commits to remove unnamed contributor

IanWood1 committed:
start on README

IanWood1 committed:
save

IanWood1 committed:
add iree command line support for pdll

IanWood1 committed:
moved pdll tests/files to subfolder and fixed cmake issues

IanWood1 committed:
clean up pdll code

IanWood1 committed:
small changes to readme

IanWood1 committed:
small changes to readme

IanWood1 committed:
readme change [skip ci]

IanWood1 committed:
add corresponding tosa tests + add comments to pdll files to match pdl

IanWood1 committed:
bazel format

IanWood1 committed:
lint script

IanWood1 committed:
fix markdownlint errors (squashed)

IanWood1 committed:
remove errant file

IanWood1 committed:
seperate pdl and pdll and move fns to utility file

IanWood1 committed:
bazel lint

IanWood1 committed:
add corresponding torch test

IanWood1 committed:
file rename

IanWood1 committed:
bazel lint

ScottTodd commented on compiler/plugins/input/TOSA/InputConversion/test/apply_pdll_patterns_tosa.mlir:
```diff
@@ -0,0 +1,54 @@
+// RUN: iree-opt --pass-pipeline="builtin.module(iree-preprocessing-apply-pdll-patterns{patterns-file=%p/tosa.pdll})" %s | FileCheck %s
+
+// CHECK-LABEL:   stream.executable private @mlp_external_f32_f32_f32_i32_i32_i32_i1_executable
```
I have a few nits about the `CHECK` lines in these test files. Mainly trying to keep the tests simple and easy to maintain. If this file was forked from another file and these issues are preexisting... sorry, that's the cost of having complicated tests and we should watch out for it spreading.

For this line, is the executable name itself important? Looks like you could drop the `f32_f32_f32_i32_i32_i32_i1` from what is being checked

MaheshRavishankar replied:
it is supposed to be some sort of mangled name. So maybe worth leaving it in as is.

qedawkins replied:
I agree with Scott's meta-point here in that this is quite an involved CHECK file. I see this same kind of thing tested across multiple input types as well. Could we deduplicate a bit?

MaheshRavishankar replied:
I like to have CHECKs for everything that is generated. So that it actually serves as a check to match what is being generated. Its just CHECK strings, some duplication on a few files seem OK?

qedawkins replied:
Torch and TOSA seem to be generating essentially the exact same verbose checks for the same rewrite, meaning we need to update it in two places if there are IR changes. It's not a huge deal but I thought the rule of thumb was to try to keep tests distinct.

ScottTodd replied:
https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-best-practices

> Tests should be minimal, and only check what is absolutely necessary.
This means that anything in the output that is not core to the functionality that you are testing should not be present in a CHECK line.

ScottTodd replied:
If you check everything that is generated, you've written a change detector test that is a chore to keep updated.

IanWood1 replied:
Would It be better to only check the call to `@mlp_external...` + the final output?

MaheshRavishankar replied:
Ok... ill defer to you guys to provide guidance here, but I dont consider lit tests to be anything more than change detector :D

qedawkins replied:
Right, they are change detectors by construction, but the idea is to make the detected changes meaningful. In this case I think it would be fine to drop either the Torch or TOSA lit test because e2e is already tested in the plugin samples for each input type.

MaheshRavishankar replied:
The Torch and TOSA lit test was just to show, you can do either, but yeah OK duplication doesnt make sense. @IanWood1  maybe just check that this is producing the dispatch. The PDL scripts are calling the same function and that is checking the generated code.

IanWood1 replied:
Okay, I removed the Torch and TOSA tests reduced the amounts of checks in plugins

qedawkins replied:
Thanks! Although I didn't mean removing the checks for all of them because it looked like it was testing the actual native rewrite pattern as well, mainly just that we didn't need the same set of verbose checks in three places. Adding back the linalg one maybe seems reasonable to me if you want.

ScottTodd commented on compiler/plugins/input/TOSA/InputConversion/test/apply_pdll_patterns_tosa.mlir:
```diff
@@ -0,0 +1,54 @@
+//  CHECK-SAME:         %[[ARG3:[a-zA-Z0-9]+]]: i32
+//  CHECK-SAME:         %[[ARG4:[a-zA-Z0-9]+]]: i32
+//  CHECK-SAME:         %[[ARG5:[a-zA-Z0-9]+]]: i32
+//  CHECK-SAME:         %[[ARG6:[a-zA-Z0-9]+]]: i1
```
Can these match `.+` (more than one of any character) instead of `[a-zA-Z0-9]+` (more than one of any lowercase letter, uppercase letter, or number)?

Same comment for `%[[STREAM0_BASE:[a-zA-Z0-9_]+]], %[[OFFSET0:[a-zA-Z0-9_]+]],`

MaheshRavishankar replied:
If the types of multiple operands are same, then `[[[.+]]]` matches more than just the immediate argument. So I also try to use `[a-zA-Z0-9_]+`  in such cases.

ScottTodd replied:
Ahhh, I see. The regex is greedy and doesn't stop after the first match (lazy)

ScottTodd commented on compiler/plugins/input/TOSA/InputConversion/test/apply_pdll_patterns_tosa.mlir:
```diff
@@ -0,0 +1,54 @@
+//   CHECK-DAG:       %[[C4:.+]] = arith.constant 4 : i32
+//   CHECK-DAG:       %[[C8:.+]] = arith.constant 8 : i32
+//   CHECK-DAG:       %[[C2:.+]] = arith.constant 2 : i32
+//   CHECK-DAG:       %[[DORELU:.+]] = arith.constant true
```
Is the "Relu" capture name important here? I don't see any references to that particular op in the input IR. Could capture as "CTRUE" or something

edit: I see relu in `tosa.pdll`, so I guess that makes sense

ScottTodd commented on compiler/plugins/input/TOSA/InputConversion/test/tosa.pdll:
Just curious - is there a syntax highlighting language definition for .pdll files? Would be nice :P

MaheshRavishankar replied:
There is. there is a mlir-pdll-lsp-server.

ScottTodd replied:
That's for https://microsoft.github.io/language-server-protocol/ which provides IDE features like auto complete, go to definition, etc. . Syntax highlighting formats are related but are usually much simpler (e.g. https://github.com/llvm/llvm-project/blob/main/mlir/utils/textmate/mlir.json).

Ah! Found https://github.com/llvm/llvm-project/blob/main/mlir/utils/vscode/pdll-grammar.json, which is included in the [MLIR VSCode extension](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-mlir). That works for me locally: 
![image](https://github.com/iree-org/iree/assets/4010439/ac25348b-832c-4cbf-972b-f5a71093ef0a)

We (the MLIR community) could also try to get that syntax/grammar added to GitHub so there would be syntax highlighting there too.

IanWood1 replied:
Github highlighting would be nice!

ScottTodd commented on compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLLPatterns.cpp:
```diff
@@ -0,0 +1,122 @@
+namespace {
+
+class ApplyPDLLPatternsPass
+    : public iree_compiler::Preprocessing::impl::ApplyPDLLPatternsPassBase<
```
nit: already in the `iree_compiler` namespace, can simplify this?
```suggestion
    : public Preprocessing::impl::ApplyPDLLPatternsPassBase<
```

I see a bunch of these `public iree_compiler::` in the Preprocessing/ folder, could clean those up too (in a separate PR to keep this one focused)

ScottTodd commented on compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLLPatterns.cpp:
```diff
@@ -0,0 +1,122 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "Utilities.h"
```
Use the full (relative to IREE compiler source root) path for includes.

```suggestion
#include "iree/compiler/Preprocessing/Common/Utilities.h"
```

ScottTodd commented on compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLPatterns.cpp:
```diff
@@ -4,22 +4,21 @@
 // See https://llvm.org/LICENSE.txt for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
+#include "Utilities.h"
```
use full include path

ScottTodd commented on compiler/src/iree/compiler/Preprocessing/Common/BUILD.bazel:
```diff
@@ -65,6 +68,10 @@ iree_compiler_cc_library(
         "@llvm-project//mlir:MemRefDialect",
         "@llvm-project//mlir:PDLDialect",
         "@llvm-project//mlir:PDLInterpDialect",
+        "@llvm-project//mlir:PDLLAST",
+        "@llvm-project//mlir:PDLLCodeGen",
+        "@llvm-project//mlir:PDLLODS",
+        "@llvm-project//mlir:PDLLParser",
```
Roughly how heavy are these new dependencies? (Build actions, binary size, etc.)

MaheshRavishankar replied:
shouldnt be too much... its just a light-weight front end

ScottTodd commented on compiler/src/iree/compiler/Pipelines/Options.cpp:
```diff
@@ -242,8 +244,11 @@ void PreprocessingOptions::bindOptions(OptionsBinder &binder) {
+  binder.opt<std::string>(
+      "iree-preprocessing-pdll-spec-filename", preprocessingPDLLSpecFilename,
+      llvm::cl::desc("File name of a PDLL spec to use for preprocessing"),
       llvm::cl::cat(category));
```
Is it possible to pass multiple spec files of each type? Should these be lists instead of singular strings?

MaheshRavishankar replied:
I think so, but for now, starting with one.

ScottTodd commented on compiler/src/iree/compiler/Preprocessing/Common/Utilities.cpp:
```diff
@@ -0,0 +1,492 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "Utilities.h"
```
use full include path

ScottTodd commented on compiler/src/iree/compiler/Preprocessing/Common/Utilities.h:
```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_PREPROCESSING_COMMON_PDLUTILITIES_H_
+#define IREE_COMPILER_PREPROCESSING_COMMON_PDLUTILITIES_H_
```
Either rename the file to PDLUtilities.h or update this header guard to match the file. When the header guard doesn't match the file name it can be too easy to accidently add a file that uses the same header guard, and then only one of the files will actually be included x_x

IanWood1 replied:
thank you, renamed it to utilities and didnt change the guard

ScottTodd commented on samples/custom_dispatch/cpu/mlp_plugin/README.md:
```diff
@@ -0,0 +1,23 @@
+In contrast with `plugin/example.mlir`, `mlp_plugin/mlp.mlir` does not
+use any external function calls, but is instead composed of only builtin
+mlir ops. iree-compile uses a rewrite pattern to match and rewrite the input MLIR, replacing matched ops with a call to `mlp_external()` defined in 
+[mlp_plugin.c](/samples/custom_dispatch/cpu/mlp_plugin/mlp_plugin.c).
```
nit: wrap at 80 (or 100) characters consistently
```suggestion
mlir ops. iree-compile uses a rewrite pattern to match and rewrite the input
MLIR, replacing matched ops with a call to `mlp_external()` defined in
[mlp_plugin.c](/samples/custom_dispatch/cpu/mlp_plugin/mlp_plugin.c).
```

ScottTodd commented on samples/custom_dispatch/cpu/mlp_plugin/pdll/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+
```
please add a copyright header to new source files (CMakeLists.txt count as source files)

```
# Copyright 2024 The IREE Authors
#
# Licensed under the Apache License v2.0 with LLVM Exceptions.
# See https://llvm.org/LICENSE.txt for license information.
# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```

ScottTodd commented on samples/custom_dispatch/cpu/mlp_plugin/pdll/mlp_linalg_pdll.mlir:
```diff
@@ -0,0 +1,114 @@
+//       CHECK:     func.func @mlp_external_entry_point
+//  CHECK-SAME:         %[[ARG0:[a-zA-Z0-9]+]]: !stream.binding
+//  CHECK-SAME:         %[[ARG1:[a-zA-Z0-9]+]]: !stream.binding
+//  CHECK-SAME:         %[[ARG2:[a-zA-Z0-9]+]]: !stream.binding
```
same comment as before about keeping lit tests simple, might be able to match `.+` here instead of `[a-zA-Z0-9]+`

ScottTodd commented on compiler/src/iree/compiler/Preprocessing/Common/CMakeLists.txt:
```diff
@@ -46,6 +49,10 @@ iree_cc_library(
+    MLIRPDLLAST
+    MLIRPDLLCodeGen
+    MLIRPDLLODS
+    MLIRPDLLParser
```
So for the `ld.lld: error: unable to find library -lMLIRPDLLParser` error on the byo_llvm job: https://github.com/iree-org/iree/actions/runs/8806295000/job/24172166753?pr=17144#step:4:15043, I suspect we'll need to add some options to https://github.com/iree-org/iree/blob/main/build_tools/llvm/byo_llvm.sh. It doesn't look like we'll need to distribute `mlir-pdll` ourselves or work that into the build/tests with how this is set up 🤞

IanWood1 replied:
@ScottTodd do you mean adding the dependencies mentioned above to [mlir_config.cmake](https://github.com/iree-org/iree/blob/26b81db6a82dd37ba865421197d26f2a0768cbd6/build_tools/llvm/mlir_config.cmake) which is used by byo_llvm.sh? Or are you thinking of something different?

ScottTodd replied:
I'm not sure. Those files are where I'd start looking though. I think you should be able to reproduce the CI failure by following what https://github.com/iree-org/iree/blob/main/build_tools/cmake/build_and_test_byo_llvm.sh does

IanWood1 replied:
My initial thought was that since we are not building `mlir-pdll`, we wouldn't be building any of its PDLL related dependencies. That turns out to not be the case. Looking at the azure logs revealed the other PDLL libraries were being built and installed, except for `MLIRPDLLParser`.

It turns out that inside of the [cmakelists](https://github.com/llvm/llvm-project/blob/ddb67e68474092435634dd8a8e1c166f6ecd0e1a/mlir/lib/Tools/PDLL/Parser/CMakeLists.txt#L2), `MLIRPDLLParser` is added as a dependency with `llvm_add_dependency` instead of the normal `mlir_add_dependency`. I'm not entirely sure why this was done, but it was changed when tablegen became a dependency. Since tablegen isn't included in the single shared library during certain builds, neither could `MLIRPDLLParser`. This issue arises because `mlir_add_dependency` also calls `add_mlir_library_install`. In fact, all other uses of `llvm_add_dependency` inside of mlir make sure to explicitly add the install library ([example](https://github.com/llvm/llvm-project/blob/ddb67e68474092435634dd8a8e1c166f6ecd0e1a/mlir/lib/Tools/mlir-tblgen/CMakeLists.txt#L4)) except for `MLIRPDLLParser`.

tl;dr The dependencies of `mlir-pdll` are already being built by the current configuration. The issue is that `MLIRPDLLParser` is not setup to be installed the llvm repo.


Is it possible to install manually it during the CI build?

IanWood1 replied:
Also, I added `add_mlir_library_install` to the llvm cmakelists to test if that would fix it, and it built the library and linked properly. However, that's probably not a viable solution

ScottTodd replied:
That sounds like something to fix upstream somehow :/

> Is it possible to install manually it during the CI build?

Maybe? The byo_llvm path is already precarious, so

1. Developers may have to follow some extra steps carefully
2. If we (and upstream LLVM/MLIR) can make that easier, we should

ScottTodd replied:
Another option is to disable that part of PDLL support in the byo_llvm case. I don't know off the top of my head how we'd do that though.

bjacob replied:
i'm out today; to unblock this and for the sake of incrementalism, i would suggest taking the easy way out for now here (disable PDLL in byo-llvm build) while filing an issue about doing The Right Thing and referencing it in a TODO comment for now.

ScottTodd reviewed (commented):
Cool! The structure of this looks good to me. Had a bunch of comments about lower level details and overall project health, then some ideas for fixing the failing CI job.

MaheshRavishankar reviewed (approved):
Apart from Scott's comments, this looks good to me. THanks!

IanWood1 committed:
requested changes

change #include to use full path, run and fix markdown lint errors in mlp_plugin/README.md, and add copyright

IanWood1 committed:
replace deprecated methods

IanWood1 committed:
remove pdll from byo builds

IanWood1 committed:
remove pdlutilities (from rebase)

MaheshRavishankar commented on build_tools/cmake/iree_llvm.cmake:
```diff
@@ -84,6 +84,10 @@ macro(iree_llvm_configure_bundled)
+
+  # TODO(#17233) 
+  set(IREE_LLVM_BUNDLED ON)
+  add_definitions(-DIREE_LLVM_BUNDLED)
```
I am not sure this is the way to do this... @ScottTodd  can you please verify/suggest a solution.

ScottTodd replied:
If the upstream change works and is uncontroversial, we should land it and always enable in this repo. Branching logic in the build system gets complicated to maintain over time. Any particular reason why https://github.com/llvm/llvm-project/pull/90411 is a draft?

IanWood1 replied:
I left it as a draft because I was getting some weirdness when linking against MLIRPDLLParser in byo_llvm (when making changes locally). For some reason, iree-compile was crashing on some testcases that were completely unrelated to pdll (just made a comment about it on #17233). I wanted to resolve that before opening the pr so I was sure the upstream changes would work

ScottTodd replied:
I'd _really_ like to avoid the `#ifdef IREE_LLVM_BUNDLED` changes here. That's going to be very difficult to untangle later if we land this as-is. What is blocking the upstream fixes?

The logs in https://gist.github.com/IanWood1/c2ea38035453b8a17b9d7c8b78fb9cd6 suggest some real issues to debug.

MaheshRavishankar commented on compiler/plugins/input/TOSA/InputConversion/test/BUILD.bazel:
```diff
@@ -17,6 +17,8 @@ iree_lit_test_suite(
         [
             "apply_pdl_patterns_tosa.mlir",
+            # TODO(#17233)
+            # "apply_pdll_patterns_tosa.mlir",
```
Huh, too bad.. would be good if we could run this tests... but I am guessing we cant "conditionally" add a test. So I am fine with this for now.

MaheshRavishankar commented on compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLLPatterns.cpp:
```diff
@@ -0,0 +1,139 @@
+// TODO(#17233)
+// Its easier to give these functions definitions than try to dissable
+// tblgen, just return failure
+#ifndef IREE_LLVM_BUNDLED
```
This is not great, but I understand why this is done this way. I am OK to land this cause this is two experimental features (byo_llvm and pdll support) clashing. It would probably be better if this wasnt built at all if `IREE_LLVM_BUNDLED` wasnt set, but I dont know if there is an alternative. I am OK with this for the time being.

MaheshRavishankar commented on compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLLPatterns.cpp:
```diff
@@ -0,0 +1,139 @@
+// tblgen, just return failure
+#ifndef IREE_LLVM_BUNDLED
+  return failure();
+#endif
```
You could just do 
```
#ifdef

#else

#endif
```

IanWood1 committed:
requested changes to preprocessor

qedawkins commented on compiler/plugins/input/TOSA/InputConversion/test/tosa.pdll:
```diff
@@ -0,0 +1,112 @@
+      reluOp, attr<"\"mlp_external\"">, 
+      inputs, repl_vals, 
+      repl_dims, other);
+  }; 
```
I see almost this exact spec, in particular with all of the same docs in `torch.pdll`, `mlp_linalg_spec.pdll`, `mlp_torch_spec.pdll`, and `mlp_tosa_spec.pdll`. This is a fairly large spec file to use to just show proof of life for different kinds of inputs. Can we deduplicate these a bit?

IanWood1 committed:
Merge branch 'main' into pdll_branch_2

IanWood1 committed:
deduplicate and slim down tests

ScottTodd reviewed (commented):
What's the status of this PR? I still see it in my review queue with some tricky `IREE_LLVM_BUNDLED` handling etc.

Do we still want/need PDLL support like this somehow?
